client/go/dogma/content_service.go

```diff
@@ -0,0 +1,320 @@
+	Detail  string `json:"detail,omitempty"`
+	Markup  string `json:"markup,omitempty"`
+}
+
+// Change represents a change to commit in the repository.
+type Change struct {
+	Path string `json:"path"`
+
+	// can be UPSERT_JSON, UPSERT_TEXT, REMOVE, RENAME, APPLY_JSON_PATCH or APPLY_TEXT_PATCH
+	Type    string      `json:"type"`
+	Content interface{} `json:"content,omitempty"`
+}
+
+func (con *contentService) listFiles(ctx context.Context,
+	projectName, repoName, revision, pathPattern string) ([]*Entry, *http.Response, error) {
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/tree%v", defaultPathPrefix, projectName, repoName, pathPattern)
+
+	if len(revision) != 0 {
+		u += "?revision=" + revision
+	}
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var entries []*Entry
+	res, err := con.client.do(ctx, req, &entries)
+	if err != nil {
+		return nil, res, err
+	}
+	return entries, res, nil
+}
+
+func (con *contentService) getFile(
+	ctx context.Context, projectName, repoName, revision string, query *Query) (*Entry, *http.Response, error) {
+	path := query.Path
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/contents%v", defaultPathPrefix, projectName, repoName, path)
+	urlQuery, err := getFileURLQuery(revision, path, query.Expressions)
+	if err != nil {
+		return nil, nil, err
+	}
+	u += urlQuery
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	entry := new(Entry)
+	res, err := con.client.do(ctx, req, entry)
+	if err != nil {
+		return nil, res, err
+	}
+
+	return entry, res, nil
+}
+
+// getFileURLQuery currently only supports JSON path.
+func getFileURLQuery(revision, path string, jsonPaths []string) (urlQuery string, err error) {
+	if len(jsonPaths) != 0 {
+		if urlQuery, err = getJSONPaths(path, jsonPaths); err != nil {
+			return "", err
+		}
+
+		if len(revision) != 0 {
+			// have both of the jsonpath and the revision
+			urlQuery += "&revision=" + revision
+		}
+	} else if len(revision) != 0 {
+		// have the revision only
+		urlQuery += "?revision=" + revision
+	}
+	return urlQuery, nil
+}
+
+func getJSONPaths(path string, jsonPaths []string) (urlQuery string, err error) {
+	if !strings.HasSuffix(strings.ToLower(path), "json") {
+		return "", fmt.Errorf("the extension of the file should be .json (path: %v)", path)
+	}
+	for i, jsonPath := range jsonPaths {
+		if i == 0 {
+			urlQuery = "?"
+		} else {
+			urlQuery += "&"
+		}
+		urlQuery += fmt.Sprintf("jsonpath=%v", jsonPath)
+	}
+	// the urlQuery starts with '?'
+	return urlQuery, nil
+}
+
+func (con *contentService) getFiles(ctx context.Context,
+	projectName, repoName, revision, pathPattern string) ([]*Entry, *http.Response, error) {
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/contents%v", defaultPathPrefix, projectName, repoName, pathPattern)
+
+	if len(revision) != 0 {
+		u += "?revision=" + revision
+	}
```

Comment:
how about use [url#Values pacakge](https://golang.org/pkg/net/url/#Values) to format

Reply:
Oops, I totally misunderstood the API. Thank you for letting me know that.